File: src/host/lib/request.ml
(**********************************************************************)
(*                                                                    *)
(*              This file is part of the RFSM package                 *)
(*                                                                    *)
(*  Copyright (c) 2018-present, Jocelyn SEROT.  All rights reserved.  *)
(*                                                                    *)
(*  This source code is licensed under the license found in the       *)
(*  LICENSE file in the root directory of this source tree.           *)
(*                                                                    *)
(**********************************************************************)

type t =
    GetVersion
  | CheckFragment of Fragment.t
  | ScanFragment of Fragment.t
  | Compile of string list (* args, including anonymous one(s) *)
  | Close
  [@@deriving show]

exception Invalid of string

let of_yojson (json : Yojson.Basic.t) : t =
  let open Yojson.Basic.Util in
  match json |> member "kind" |> to_string with
  | "version" -> GetVersion
  | "close" -> Close
  | "check" ->
      let frag = Fragment.of_json (json |> member "fragment") in
      CheckFragment frag
  | "scan" ->
      let frag = Fragment.of_json (json |> member "fragment") in
      ScanFragment frag
  | "compile" ->
      let files = json |> member "args" |> to_list |> List.map to_string in
      Compile files
  | other ->
      
      raise (Invalid other)

let of_string s = 
  s |> Yojson.Basic.from_string |> of_yojson

let to_json (r : t) : Yojson.Basic.t =
  match r with
  | GetVersion ->
      `Assoc [("kind", `String "version")]
  | Close ->
      `Assoc [("kind", `String "close")]
  | CheckFragment frag ->
      `Assoc [
        ("kind", `String "check");
        ("fragment", Fragment.to_json frag)
      ]
  | ScanFragment frag ->
      `Assoc [
        ("kind", `String "scan");
        ("fragment", Fragment.to_json frag)
      ]
  | Compile files ->
      `Assoc [
        ("kind", `String "compile");
        ("args", `List (List.map (fun s -> `String s) files))
      ]

let to_string r =
  Yojson.Basic.pretty_to_string (to_json r)

File: src/guests/std/tests/server/client/request.cpp
/***********************************************************************/
/*                                                                     */
/*       This file is part of the Grasp software package               */
/*                                                                     */
/*  Copyright (c) 2019-present, Jocelyn SEROT (jocelyn.serot@uca.fr)   */
/*                       All rights reserved.                          */
/*                                                                     */
/*    This source code is licensed under the license found in the      */
/*      LICENSE file in the root directory of this source tree.        */
/*                                                                     */
/***********************************************************************/

#include "request.h"

Request Request::GetVersion() { return Request(Kind::GetVersion); }
Request Request::CheckFragment(const Fragment &frag) { return Request(frag); }
Request Request::Compile(const QStringList &args) { return Request(args); }
Request Request::Close() { return Request(Kind::Close); }

Request::Kind Request::kind() const { return m_kind; }
Fragment Request::fragment() const { return m_fragment; }
QStringList Request::files() const { return m_files; }

QJsonObject Request::toJson() const {
    QJsonObject obj;
    switch (m_kind) {
    case Kind::GetVersion:
        obj["kind"] = "version";
        break;
    case Kind::Close:
        obj["kind"] = "close";
        break;
    case Kind::CheckFragment:
        obj["kind"] = "check";
        obj["fragment"] = m_fragment.toJson();
        break;
    case Kind::Compile: {
        obj["kind"] = "compile";
        QJsonArray arr;
        for (const auto &s : m_files) {
            arr.append(s);
        }
        obj["args"] = arr;
        break;
    }
    }
    return obj;
}

Request Request::fromJson(const QJsonObject &obj) {
    QString kind = obj["kind"].toString();
    if (kind == "GetVersion") {
        return GetVersion();
    } else if (kind == "Close") {
        return Close();
    } else if (kind == "CheckFragment") {
        Fragment f = Fragment::fromJson(obj["fragment"].toObject());
        return CheckFragment(f);
    } else if (kind == "Compile") {
        QStringList files;
        for (const auto &v : obj["files"].toArray()) {
            files.append(v.toString());
        }
        return Compile(files);
    }
    return Close(); // default fallback
}

Request Request::fromString(const QString &jsonStr)
{
    QJsonParseError err;
    QJsonDocument doc = QJsonDocument::fromJson(jsonStr.toUtf8(), &err);
    if (err.error != QJsonParseError::NoError) {
        return Request::Close(); // default fallback
    }
    return fromJson(doc.object());
}

Request::Request(Kind kind) : m_kind(kind) {}
Request::Request(const Fragment &frag) : m_kind(Kind::CheckFragment), m_fragment(frag) {}
Request::Request(const QStringList &files) : m_kind(Kind::Compile), m_files(files) {}
